streams/src/commonMain/kotlin/com/mirego/trikot/streams/reactive/promise/Promise.kt

```diff
@@ -0,0 +1,144 @@
+import com.mirego.trikot.foundation.concurrent.dispatchQueue.SynchronousSerialQueue
+import com.mirego.trikot.streams.cancellable.CancellableManager
+import com.mirego.trikot.streams.reactive.BehaviorSubjectImpl
+import com.mirego.trikot.streams.reactive.Publishers
+import com.mirego.trikot.streams.reactive.observeOn
+import com.mirego.trikot.streams.reactive.subscribe
+import org.reactivestreams.Publisher
+import org.reactivestreams.Subscriber
+
+class Promise<T> internal constructor(
+    upstream: Publisher<T>,
+    cancellableManager: CancellableManager? = null
+) : Publisher<T> {
+
+    private val result = BehaviorSubjectImpl<T>()
+    private val serialQueue = SynchronousSerialQueue()
+
+    private val isCancelled: AtomicReference<Boolean> = AtomicReference(false)
+    private val internalCancellableManager: CancellableManager = CancellableManager().also {
+        cancellableManager?.add(it)
+    }
+
+    init {
+        upstream
+            .observeOn(serialQueue)
+            .subscribe(internalCancellableManager,
+                onNext = { value ->
+                    if (!isCancelled.value) {
+                        result.value = value
+                        result.complete()
+
+                        internalCancellableManager.cancel()
+                    }
+                },
+                onError = { error ->
+                    if (!isCancelled.value) {
+                        result.error = error
+
+                        internalCancellableManager.cancel()
+                    }
+                },
+                onCompleted = {
+                    if (!isCancelled.value) {
+                        if (result.value == null && result.error == null) {
+                            result.error = EmptyPromiseException
+                        }
+                        internalCancellableManager.cancel()
+                    }
+                }
+            )
+
+        internalCancellableManager.add {
+            serialQueue.dispatch {
+                isCancelled.setOrThrow(false, true)
+
+                if (result.value == null && result.error == null) {
+                    result.error = CancelledPromiseException
+                }
+            }
+        }
+    }
+
+    override fun subscribe(s: Subscriber<in T>) {
+        result.subscribe(s)
+    }
+
+    fun onSuccess(accept: (T) -> Unit): Promise<T> = then(
+        onSuccess = { accept(it); resolve(it) },
+        onError = ::reject
+    )
+
+    fun <R> onSuccessReturn(apply: (T) -> Promise<R>): Promise<R> = then(
+        onSuccess = apply,
+        onError = ::reject
+    )
+
+    fun onError(accept: (Throwable) -> Unit): Promise<T> = then(
+        onSuccess = ::resolve,
+        onError = { accept(it); reject(it) }
+    )
+
+    fun onErrorReturn(apply: (Throwable) -> Promise<T>): Promise<T> = then(
+        onSuccess = ::resolve,
+        onError = apply
+    )
+
+    fun finally(execute: () -> Unit): Publisher<T> = then(
```

Comment:
👏 Nice un finally

Reply:
I also made sure that we break the "promise chain" by returning a Publisher instead of a Promise for this operator! This way, we can't call :

```
aPromise // Promise<T>
    .onSuccess { ... } // Promise<T>
    .onError { ... } // Promise<T>
    .finally { ... } // Publisher<T>
    .onSuccess { ... } // <-- Unresolved reference
```